Adds Lightning Bolt spell

AnturK opened the pull request:
### Lightning Bolt

Default Cooldown : 30s
Tap once to start charging
Tap second time to release lightning
Depending on how long you charged different results :

| Charge Time | Effect |
| --- | --- |
| 0-2.5s | Minor shock 10 burn + usual electrocution stun |
| 2.5-7.5s | Major shock 25 burn + usual electrocution stun |
| 7.5-10s | Chain lightning bouncing up to 6 times dealing the 25 burn and stun |
| 10s+ | Electrocutes caster for 25 burn damage and the stun |

![bolt](https://cloud.githubusercontent.com/assets/4047233/6813039/c977679a-d272-11e4-919a-944591dc2ee1.png)

Because i enjoy murder wizards. Tried to make the beam code use transform and animate but either there's some unstated restrictions on them i'm not aware of or they are bugged.

AnturK committed:
Adds Lightning Bolt spell

Dorsisdwarf commented on code/datums/spells/lightning.dm:
```diff
@@ -0,0 +1,89 @@
+	charge_type = "recharge"
+	charge_max	= 300
+	clothes_req = 1
+	invocation = "UN'LTD P'WAH!"
```
Not magicky enough, sounds like a whispered cult spell.

If you disagree, post vocaroo of you screaming this at the top of your lungs while running.

vista- replied:
@Dorsisdwarf How about using random wizard-ized (the apostrophes) swearwords?
https://www.youtube.com/watch?v=X_d4j3UlZ18

Razharas commented:
http://www.youtube.com/watch?v=j_ekugPKqFw

AnturK commented:
http://img.4plebs.org/boards/tv/image/1408/66/1408664705093.webm

AnturK commented:
I was briefly considering adding this to Sith Lord bundle but meh

hornygranny commented:
awesome

GunHog commented:
Lightning Bolt! Lightning Bolt! Lightning Bolt! Lightning Bolt! Lightning Bolt! Lightning Bolt! Lightning Bolt! :+1:

Dorsisdwarf commented:
Oh hell yes replace the lame force power with LIGHTNINGBOLT. Maybe remove their noslips if the pack still has 'em

d3athrow commented:
THEFT

hornygranny commented:
called the police, this is an outrage

An0n3 commented:
Needs audio from that shitty youtube video to complete the effect

Incoming5643 commented:
Hooray now I don't have to do it!

I don't think it needs a blowback effect for overcharging though. Wizards got it hard enough without having to have to predict byond time passage. That or at least still do the chain lightning effect in addiction to the self harm.

Incoming5643 commented on code/datums/spells/lightning.dm:
```diff
@@ -0,0 +1,89 @@
+	invocation = "UN'LTD P'WAH!"
+	invocation_type = "shout"
+	range = 7
+	cooldown_min = 30
```
30 seconds at x1
23.3 seconds at x2
16.5 seconds at x3
9.8 seconds at x4
3 seconds at x5

Is it safe to cast this when the cooldown is shorter than the max charge time?
Do you want it to ramp that efficiently?

AnturK commented:
Yeah,  it works for cooldowns below 10 seconds,the actual cooldown starts at the discharge so there's at least that 3s delay between shocks. Might use a bit of a longer delay i guess bu then again that's 5 spell levels

Man-Shroom commented:
Boooo on the self damage. Adding another way for lag to kill people is bad

Iamgoofball commented:
:-1: on the self damage and overcharge and self use downsides, no other wiz spells have this

:+1: on everything else

Xhuis commented:
It'd be nice if someone wearing a hardsuit was more effective but I realize that'd probably be a bitch to code. The spell itself is cool.

Lumipharon commented:
The self rape for overtiming is pretty bad, byond time isn't consistant.
Making it chain enemies as normal AND effect the wizard would be alright though.

Also yes, the sith bundle would actually be neat if it got this, although probably a bit on the strong side.

RemieRichards commented on code/datums/spells/lightning.dm:
```diff
@@ -0,0 +1,89 @@
+		user << "<span class='notice'>They are too far away!</span>"
+		return
+
+	user.Beam(target,icon_state="lightning",icon='icons/effects/effects.dmi',time=5)
```
You should always use get_line() with Beam()
to ensure the beam doesn't go through walls.

RemieRichards commented on code/datums/spells/lightning.dm:
```diff
@@ -0,0 +1,89 @@
+	energy = 0
+
+/obj/effect/proc_holder/spell/targeted/lightning/proc/Bolt(mob/origin,mob/target,bolt_energy,mob/user = usr)
+	origin.Beam(target,icon_state="lightning",icon='icons/effects/effects.dmi',time=5)
```
You should always use get_line() with Beam()
to ensure the beam doesn't go through walls.

AlexCrimson commented:
Awesome. Should turn this into a gun too.

Giacom commented on code/datums/spells/lightning.dm:
```diff
@@ -0,0 +1,89 @@
+/obj/effect/proc_holder/spell/targeted/lightning/proc/StartChargeup(mob/user = usr)
+	ready = 1
+	user << "<span class='notice'>You start gathering the power.</span>"
+	//TODO: Add visual indictaor of charging
```
I think a good visual indicator would be to add a glow to the caster, like with Cold Resistance. It could pulse quicker the longer you charge it.

Giacom commented on code/datums/spells/lightning.dm:
```diff
@@ -0,0 +1,89 @@
+		user << "<span class='notice'>No target found in range.</span>"
+		return
+
+	var/mob/living/carbon/target = targets[1]
```
Do you get to choose the target or does it choose someone around the player? If it is the latter, maybe it should always shoot the closest target.

RemieRichards replied:
It's random, since ```random_target``` was also added (and set on this spell) in this PR.

Also yeah closest target should work.
```
var/mob/living/carbon/target
for(var/mob/living/L in targets)
    if(target && (get_dist(user,L) < get_dist(user,target)))
        target = L
    else
        target = L
```

AnturK replied:
Good idea will swap that

RemieRichards commented on code/datums/spells/lightning.dm:
```diff
@@ -0,0 +1,89 @@
+
+	var/mob/living/carbon/target = targets[1]
+
+	if(!(target in oview(range)))
```
You can swap this for a distance check ```get_dist(user, target) > 7``` (Using the value for view instead of 7, I just can't recall it atm) since the target has to have been in view in the first place to even be in the targets list, plus view()+oview() contain everything you can see, which is worse than just a range check.

RemieRichards commented on code/datums/spells/lightning.dm:
```diff
@@ -0,0 +1,89 @@
+	var/energy = 0
+	var/ready = 0
+
+/obj/effect/bolt
```
Correct me if If I'm wrong, but this is unused?

AnturK replied:
Ah right,  forgot to remove this.  Leftover from transform using version

LaharlMontogmmery commented:
When I saw this, I'll be honest, I thought : 

"This lightning bounce code could totally be repurposed for a possible Tesla engine/singularity."

We could use lightning bolts for when someone is just hit by a high electrical charge (doors shocked with singulo wired into the grid, wink wink.) . Even better, a bloody Tesla gun.

Just rambling with silly ideas here.

AlexCrimson commented:
Yes please a Tesla gun. Along with Tesla Trooper armor. We need more Command & Conquer stuff.

...then you can add a Desolator Rad cannon.

GunHog commented:
I very much would LOVE to code a Tesla gun. You could remove the overcharge penalty from the Wizard spell and give it to the gun instead.

xxalpha commented:
Make it call emp_act on user contents.

AnturK committed:
- adds LoS obstacle check
- some sanity checks
- Removes the blowback, now the spell just fizzles harmlessly

Jordie0608 commented:
Changelog

Dorsisdwarf commented:
The main reason everyone was so sad about the Tesla engine failure was that it would have provided a framework for LIGHTNINGBOLT, but I guess this could work backwards?

AnturK committed:
- Adds simple charging visual
- Adds sound effect
- Changelog

AnturK commented:
Used the defib sound since it sounds bit more zappy than the sparks ones and i couldn't find nice thunderclap ogg.
